core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ImplementTableFunctionSource.java

```diff
@@ -318,7 +313,9 @@ private static NodeWithSymbols appendHelperSymbolsForSource(
                 false);
 
         // If the source has set semantics, its specification is present, even if there is no partitioning or ordering specified.
-        DataOrganizationSpecification specification = argumentProperties.getSpecification().orElseThrow();
+        // If the source has row semantics, its specification is empty. Currently, such source is processed
+        // as if it was a single partition. Alternatively, it could be split into smaller partitions of arbitrary size.
+        DataOrganizationSpecification specification = argumentProperties.getSpecification().orElse(new DataOrganizationSpecification(ImmutableList.of(), Optional.empty()));
```

Comment:
Create a constant for `new DataOrganizationSpecification(ImmutableList.of(), Optional.empty())` for clarity. You can name it `UNORDERED_SINGLE_PARTITION`